File: web/src/state/watched-exchange.ts
/**
 * `watched` on the sync record: exporting both a live mark and its removal,
 * and taking either back in.
 *
 * Kept out of `store.ts` for the same reason `lists-exchange.ts` is: the
 * shape here is the one with a tombstone, and turning the `removed_at`
 * column into a row for the wire — and a row off the wire back into that
 * column — is one job, not a pair of `importMerged` branches.
 *
 * A live mark and its removal are exclusive by the time they reach here:
 * `watched-reconcile.ts` has already decided, per title, which one a merge
 * agreed on. `importWatched` therefore stays a single-kind writer, the same
 * shape `import_progress` already is — no branch inside the loop deciding
 * which fact a row represents.
 */

import type { Database } from "bun:sqlite";
import type { UnwatchedRow, WatchedRow } from "./sync-record";

interface StandingWatched {
  updatedAt: number;
  removedAt: number | null;
}

function standingFor(db: Database, profileId: string, setId: string): StandingWatched | null {
  return db
    .query(
      "SELECT removed_at AS removedAt, CASE WHEN removed_at IS NOT NULL THEN removed_at ELSE finished_at END AS updatedAt " +
        "FROM watched WHERE profile_id = ?1 AND set_id = ?2",
    )
    .get(profileId, setId) as StandingWatched | null;
}

/** Every `watched` row this profile has, live and removed, for the wire. */
export function exportWatched(
  db: Database | null,
  profileId: string,
): { watched: WatchedRow[]; unwatched: UnwatchedRow[] } {
  if (!db) return { watched: [], unwatched: [] };
  const rows = db
    .query(
      "SELECT set_id AS setId, finished_at AS finishedAt, removed_at AS removedAt FROM watched WHERE profile_id = ?1",
    )
    .all(profileId) as { setId: string; finishedAt: number; removedAt: number | null }[];

  const watched: WatchedRow[] = [];
  const unwatched: UnwatchedRow[] = [];
  for (const row of rows) {
    if (row.removedAt !== null) {
      unwatched.push({ setId: row.setId, updatedAt: row.removedAt, lastFinishedAt: row.finishedAt });
    } else {
      watched.push({ setId: row.setId, updatedAt: row.finishedAt });
    }
  }
  return { watched, unwatched };
}

/**
 * A completion supersedes a stale local position, whether or not it is
 * news: a device that already knew of it may still hold a position another
 * device has only now reported. Applied unconditionally, before the
 * standing check below decides whether the mark itself is written — the
 * position's fate and the mark's are not the same question.
 */
export function importWatched(db: Database | null, profileId: string, rows: WatchedRow[]): number {
  if (!db) return 0;
  let changed = 0;
  for (const row of rows) {
    changed += db
      .query("DELETE FROM progress WHERE profile_id = ?1 AND set_id = ?2 AND updated_at <= ?3")
      .run(profileId, row.setId, row.updatedAt).changes;

    const standing = standingFor(db, profileId, row.setId);
    if (standing !== null && standing.updatedAt >= row.updatedAt) continue;
    db.query(
      `INSERT INTO watched(profile_id, set_id, finished_at, removed_at) VALUES (?1, ?2, ?3, NULL)
         ON CONFLICT(profile_id, set_id) DO UPDATE SET finished_at = excluded.finished_at, removed_at = NULL`,
    ).run(profileId, row.setId, row.updatedAt);
    changed += 1;
  }
  return changed;
}

/**
 * `lastFinishedAt` is the completion this removal took the mark from, so it
 * supersedes a stale local position the same way a live completion's own
 * time would — `merge.ts` already decided a rewatch made since survives.
 */
export function importUnwatched(db: Database | null, profileId: string, rows: UnwatchedRow[]): number {
  if (!db) return 0;
  let changed = 0;
  for (const row of rows) {
    changed += db
      .query("DELETE FROM progress WHERE profile_id = ?1 AND set_id = ?2 AND updated_at <= ?3")
      .run(profileId, row.setId, row.lastFinishedAt).changes;

    const standing = standingFor(db, profileId, row.setId);
    // `watched-reconcile.ts` gives an exact tie to the removal, not a
    // device-id tie-break, whenever the standing fact is a *live* mark — so
    // import must skip one only when that live mark is strictly newer.
    // Standing already a removal is the ordinary same-kind case, where a
    // tie changes nothing either way.
    const outrun = standing !== null &&
      (standing.removedAt !== null ? standing.updatedAt >= row.updatedAt : standing.updatedAt > row.updatedAt);
    if (outrun) continue;
    db.query(
      `INSERT INTO watched(profile_id, set_id, finished_at, removed_at) VALUES (?1, ?2, ?3, ?4)
         ON CONFLICT(profile_id, set_id) DO UPDATE SET finished_at = excluded.finished_at, removed_at = excluded.removed_at`,
    ).run(profileId, row.setId, row.lastFinishedAt, row.updatedAt);
    changed += 1;
  }
  return changed;
}

File: crates/mediagram-core/src/state/watched_exchange.rs
//! `watched` on the sync record: exporting both a live mark and its
//! removal, and taking either back in. Matches
//! `web/src/state/watched-exchange.ts`.
//!
//! A live mark and its removal are exclusive by the time they reach here:
//! `merge::watched::reconcile` has already decided, per title, which one a
//! merge agreed on. `import_watched`/`import_unwatched` therefore stay
//! single-kind writers, the same shape `import_progress` already is.

use rusqlite::{Connection, OptionalExtension, params};

use super::record::{UnwatchedRow, WatchedRow};

struct Standing {
    updated_at: f64,
    removed: bool,
}

fn standing_for(
    conn: &Connection,
    profile_id: &str,
    set_id: &str,
) -> rusqlite::Result<Option<Standing>> {
    let row: Option<(i64, Option<i64>)> = conn
        .query_row(
            "SELECT finished_at, removed_at FROM watched WHERE profile_id = ?1 AND set_id = ?2",
            params![profile_id, set_id],
            |r| Ok((r.get(0)?, r.get(1)?)),
        )
        .optional()?;
    Ok(row.map(|(finished_at, removed_at)| Standing {
        updated_at: removed_at.unwrap_or(finished_at) as f64,
        removed: removed_at.is_some(),
    }))
}

/// Every `watched` row this profile has, live and removed, for the wire.
pub fn export_watched(
    conn: &Connection,
    profile_id: &str,
) -> rusqlite::Result<(Vec<WatchedRow>, Vec<UnwatchedRow>)> {
    let mut stmt = conn.prepare(
        "SELECT set_id, finished_at, removed_at FROM watched WHERE profile_id = ?1",
    )?;
    let rows: Vec<(String, i64, Option<i64>)> = stmt
        .query_map([profile_id], |row| {
            Ok((row.get(0)?, row.get(1)?, row.get(2)?))
        })?
        .collect::<rusqlite::Result<_>>()?;

    let mut watched = Vec::new();
    let mut unwatched = Vec::new();
    for (set_id, finished_at, removed_at) in rows {
        match removed_at {
            Some(removed_at) => unwatched.push(UnwatchedRow {
                set_id,
                updated_at: removed_at as f64,
                last_finished_at: finished_at as f64,
            }),
            None => watched.push(WatchedRow {
                set_id,
                updated_at: finished_at as f64,
            }),
        }
    }
    Ok((watched, unwatched))
}

/// A completion supersedes a stale local position whether or not it is
/// news: a device that already knew of it may still hold a position
/// another device has only now reported. Applied unconditionally, before
/// the standing check decides whether the mark itself is written.
pub fn import_watched(conn: &Connection, profile_id: &str, rows: &[WatchedRow]) -> rusqlite::Result<u64> {
    let mut changed = 0u64;
    for row in rows {
        changed += conn.execute(
            "DELETE FROM progress WHERE profile_id = ?1 AND set_id = ?2 AND updated_at <= ?3",
            params![profile_id, row.set_id, row.updated_at as i64],
        )? as u64;

        let standing = standing_for(conn, profile_id, &row.set_id)?;
        if standing.is_some_and(|s| s.updated_at >= row.updated_at) {
            continue;
        }
        conn.execute(
            "INSERT INTO watched(profile_id, set_id, finished_at, removed_at) VALUES (?1, ?2, ?3, NULL)
               ON CONFLICT(profile_id, set_id) DO UPDATE SET finished_at = excluded.finished_at, removed_at = NULL",
            params![profile_id, row.set_id, row.updated_at as i64],
        )?;
        changed += 1;
    }
    Ok(changed)
}

/// `last_finished_at` is the completion this removal took the mark from,
/// so it supersedes a stale local position the same way a live
/// completion's own time would — `merge.rs` already decided a rewatch made
/// since survives.
pub fn import_unwatched(
    conn: &Connection,
    profile_id: &str,
    rows: &[UnwatchedRow],
) -> rusqlite::Result<u64> {
    let mut changed = 0u64;
    for row in rows {
        changed += conn.execute(
            "DELETE FROM progress WHERE profile_id = ?1 AND set_id = ?2 AND updated_at <= ?3",
            params![profile_id, row.set_id, row.last_finished_at as i64],
        )? as u64;

        let standing = standing_for(conn, profile_id, &row.set_id)?;
        // `merge::watched::reconcile` gives an exact tie to the removal, not
        // a device-id tie-break, whenever the standing fact is a *live*
        // mark — so this must skip one only when that live mark is strictly
        // newer. Standing already a removal is the ordinary same-kind case,
        // where a tie changes nothing either way.
        let outrun = standing.is_some_and(|s| {
            if s.removed {
                s.updated_at >= row.updated_at
            } else {
                s.updated_at > row.updated_at
            }
        });
        if outrun {
            continue;
        }
        conn.execute(
            "INSERT INTO watched(profile_id, set_id, finished_at, removed_at) VALUES (?1, ?2, ?3, ?4)
               ON CONFLICT(profile_id, set_id) DO UPDATE SET finished_at = excluded.finished_at, removed_at = excluded.removed_at",
            params![
                profile_id,
                row.set_id,
                row.last_finished_at as i64,
                row.updated_at as i64
            ],
        )?;
        changed += 1;
    }
    Ok(changed)
}

#[cfg(test)]
#[path = "watched_exchange_tests.rs"]
mod tests;
